### PR TITLE
ui: Reduce WebPack / Warning Spew during builds

### PR DESCRIPTION
**Pull request by mrtracy**
Commit makes a few changes to reduce the significant output spew from
our frontend build processes.

Fixes all React 0.15 deprecation warnings, related to usage of certain
React features which had been moved to new packages:

+ Updated several third-party react-related packages.
+ Install 'prop-types' and 'react-test-renderer' packages as
dependencies.
+ Fix all usages of `React.PropTypes` in Admin UI code.
+ Change all imports of style `import * as React from 'react'` to
`import React from 'react'`. This is necessary to prevent the
"PropTypes" deprecation message.

New settings added for build tools:

+ Begins using Webpack's 'stats' configuration value to reduce the
non-error output from webpack. The output that remains is also now
colorized, Improving readability.
+ Use the "noinfo" setting for Karma's webpackMiddleware plugin,
completely suppressing informational webpack messages while running
tests.

Makefile changes:

+ Changes "test-debug" to "test-watch". This target runs karma in a
persistent session which watches for filesystem changes and re-runs
tests whenever anything changes. This target no longer uses the Chrome
browser; the "test-debug" target was originally created because
Phantomjs was not outputting sufficient information for test failures,
but that is no longer an issue with jsdom.

### CODE REVIEW

**Comment by tamird** [comment]
> Install 'prop-types' and 'react-test-renderer' packages as
> dependencies.

prop-types is used directly; what is react-test-renderer for?

:lgtm: aside from prop-types, which seems like something we should just remove.

---

Reviewed 52 of 52 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/ui/Makefile, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYW9zlq0FetX33pgFY:-KkYW9zlq0FetX33pgFZ:bmi8ppe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/Makefile#L46)):*
> ```Makefile
> .PHONY: test-watch
> test-watch: $(YARN_INSTALLED_TARGET) | protos
> 	karma start --no-single-run --auto-watch
> ```

:100: 

---

*[pkg/ui/package.json, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYWV1RdsTQuAYqmewT:-KkYWV1SF_uxvLjUEzyw:b39uttf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/package.json#L70)):*
> ```JSON
>     "nib": "^1.1.2",
>     "nvd3": "^1.8.5",
>     "prop-types": "^15.5.10",
> ```

hm, do we need this? seems like it provides run-time type checking, which we shouldn't need because of TS.

---

*[pkg/ui/webpack.config.js, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYX0_0LWvgBTQF4S2O:-KkYX0_1bQ7TT_MX3dei:bsaj356) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/webpack.config.js#L24)):*
> ```JavaScript
> 
> // https://webpack.js.org/configuration/stats/
> const stats = {
> ```

this can just inline below, right? not sure why this deviates from the local convention.

---

*[pkg/ui/webpack.config.js, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXMOCwbc8ab7xXRz4:-KkYXMOCwbc8ab7xXRz5:bkcz8j5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/webpack.config.js#L27)):*
> ```JavaScript
>   colors: true,
>   chunks: false,
>   chunkModules: false,
> ```

are these two needed given that `chunks: false` is above? my reading of the config page indicates that `chunks: false` would turn this bit of verbosity off completely.

---

*[pkg/ui/src/components/listLink.tsx, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXYDPPZ7Bd0UDwuWL:-KkYXYDPPZ7Bd0UDwuWM:b-nburms) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/components/listLink.tsx#L26)):*
> ```XML
> export class ListLink extends React.Component<LinkProps, {}> {
>   static contextTypes = {
>     router: PropTypes.object,
> ```

what's the point of this? this seems equivalent to `{}` to me, yet I think we can probably give it a better type given that we have typings for react-router.

---

*[pkg/ui/src/containers/nodes.tsx, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXtnzAxrMF9Ae5Rir:-KkYXtnzAxrMF9Ae5Ris:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/containers/nodes.tsx#L50)):*
> ```XML
>   // See https://github.com/ReactTraining/react-router/issues/975
>   static contextTypes = {
>     router: PropTypes.object.isRequired,
> ```

ditto

---

*[pkg/ui/src/containers/databases/databases.tsx, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXx798ZByhT_JoRD-:-KkYXx798ZByhT_JoRD0:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/containers/databases/databases.tsx#L36)):*
> ```XML
>   // See https://github.com/ReactTraining/react-router/issues/975
>   static contextTypes = {
>     router: PropTypes.object.isRequired,
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-:-KkYWlgOUVXcs6UpJ_rm:b-373fww)*

**Comment by mrtracy** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/ui/package.json, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYWV1RdsTQuAYqmewT:-KklFx0F3PWUG53egVxE:bp8einy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/package.json#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hm, do we need this? seems like it provides run-time type checking, which we shouldn't need because of TS.
</blockquote></details>

Enzyme needs it.  We need to explicitly include both this and "react-test-renderer" in our dependencies because they are some kind of "soft dependency" for Enzyme; basically, enzyme doesn't require these explicitly, but will use them if they are available. However, the fallback used by Enzyme displays the deprecation warning from React.

---

*[pkg/ui/webpack.config.js, line 24 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYX0_0LWvgBTQF4S2O:-KklGDFrEQN0aBztFqV5:b5izexi) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/webpack.config.js#L24)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this can just inline below, right? not sure why this deviates from the local convention.
</blockquote></details>

Sorry, this "stats" structure can be placed in several sub-configurations; for example, in the karma config I have to move it to the "webpackMiddleware" configuration for it to be respected.  This is also the case for a configuration section called "devServer", which (when present) is placed directly in the webpack config.  So, I had originally defined the stats object up here and assigned it to both the top-level "stats" field *and* "devServer.stats".  However, I then discovered out we aren't using the devServer anywhere, so I removed that configuration section.

tldr; Done.

---

*[pkg/ui/webpack.config.js, line 27 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXMOCwbc8ab7xXRz4:-KklHkqB2pRnhllZh5m6:b-vipgtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/webpack.config.js#L27)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

are these two needed given that `chunks: false` is above? my reading of the config page indicates that `chunks: false` would turn this bit of verbosity off completely.
</blockquote></details>

You are correct, done.

---

*[pkg/ui/src/components/listLink.tsx, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXYDPPZ7Bd0UDwuWL:-KklJ7J5xOPk6nFfgR3c:bfbelqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/components/listLink.tsx#L26)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what's the point of this? this seems equivalent to `{}` to me, yet I think we can probably give it a better type given that we have typings for react-router.
</blockquote></details>

This isn't used to provide any extra compile-time or run-time type information, it's just required by react in order for a child to receive a context from a parent.  Basically, the parent needs to provide a "childContextTypes" field which describes the contexts it will provide, and children need to define this "contextTypes" field to define which contexts they accept.  React will not pass a context to a child that does not request it with this field.

---

*[pkg/ui/src/containers/nodes.tsx, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXtnzAxrMF9Ae5Rir:-KklK64IiGM1liiBKw8s:b-fvdr1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/containers/nodes.tsx#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

answered above

---

*[pkg/ui/src/containers/databases/databases.tsx, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXx798ZByhT_JoRD-:-KklK8Edqo-tB1hBwMEq:b-fvdr1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/containers/databases/databases.tsx#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

answered above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039)*

**Comment by mrtracy** [comment]
As mentioned in more detail below, PropTypes is needed because we are using react contexts to get information provided by react-router, and we must declare the contextType field on components that need this information or they will not receive it.

---

Review status: 50 of 52 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-:-KklOP8-sqGpwj9KngKz:bskqggo)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ui/package.json, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYWV1RdsTQuAYqmewT:-KklSuyJwc4POx9zEB_a:b-57m1bc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/package.json#L70)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Enzyme needs it.  We need to explicitly include both this and "react-test-renderer" in our dependencies because they are some kind of "soft dependency" for Enzyme; basically, enzyme doesn't require these explicitly, but will use them if they are available. However, the fallback used by Enzyme displays the deprecation warning from React.
</blockquote></details>

Is there an upstream issue we can point to in a TODO?

---

*[pkg/ui/src/components/listLink.tsx, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXYDPPZ7Bd0UDwuWL:-KklUMXLdqZK2wzntCil:b-ntt28p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/components/listLink.tsx#L26)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This isn't used to provide any extra compile-time or run-time type information, it's just required by react in order for a child to receive a context from a parent.  Basically, the parent needs to provide a "childContextTypes" field which describes the contexts it will provide, and children need to define this "contextTypes" field to define which contexts they accept.  React will not pass a context to a child that does not request it with this field.
</blockquote></details>

This one is different from the other two; the other two use
```
  // Magic to add react router to the context.
  // See https://github.com/ReactTraining/react-router/issues/975
  static contextTypes = {
    router: PropTypes.object.isRequired,
  }
```

Can we at least make them all the same?

Also, it looks like the referenced issue now talks about [`withRouter`](https://reacttraining.com/react-router/web/api/withRouter), which might be what we want here instead?

---

*[pkg/ui/src/containers/nodes.tsx, line 52 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KklUhB6KjFvR1i21NPn:-KklUhB6KjFvR1i21NPo:bj328ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebe4be9310c8ffbe238d16cd06ead0c09001e79/pkg/ui/src/containers/nodes.tsx#L52)):*
> ```XML
>     router: PropTypes.object.isRequired,
>   };
>   context: { router?: InjectedRouter & RouterState; };
> ```

given the `isRequired` above, I think this probably shouldn't be optional here.

---

*[pkg/ui/src/containers/databases/databases.tsx, line 38 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KklUmaadj-tbq7fLVun:-KklUmaadj-tbq7fLVuo:bj328ro) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebe4be9310c8ffbe238d16cd06ead0c09001e79/pkg/ui/src/containers/databases/databases.tsx#L38)):*
> ```XML
>     router: PropTypes.object.isRequired,
>   };
>   context: { router?: InjectedRouter & RouterState; };
> ```

given the `isRequired` above, I think this probably shouldn't be optional here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039)*

**Comment by mrtracy** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ui/package.json, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYWV1RdsTQuAYqmewT:-Kkl_g5J7dkl7AHd_avB:buro7ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/package.json#L70)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Is there an upstream issue we can point to in a TODO?
</blockquote></details>

No.

After research, react-test-renderer is an implicit dependency because the enzyme package works with multiple versions of React.  It will detect (at runtime) if you are using certain react versions. It will actually raise an error if you are using 15.5 but don't have this dependency. In the case of 15.4 it just allows the deprecation warning provided by React to occur.

Eventually they may make this an explicit dependency, in which case we will no longer need to make it implicit; however, I personally feel this is a fairly minor housekeeping concern. If Enzyme stops using this library for some reason and nothing else depends on it, the code will no longer be included our webpack builds.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ui/package.json, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYWV1RdsTQuAYqmewT:-KklckDLuhylb_vwLoMu:brfq08a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/package.json#L70)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

No.

After research, react-test-renderer is an implicit dependency because the enzyme package works with multiple versions of React.  It will detect (at runtime) if you are using certain react versions. It will actually raise an error if you are using 15.5 but don't have this dependency. In the case of 15.4 it just allows the deprecation warning provided by React to occur.

Eventually they may make this an explicit dependency, in which case we will no longer need to make it implicit; however, I personally feel this is a fairly minor housekeeping concern. If Enzyme stops using this library for some reason and nothing else depends on it, the code will no longer be included our webpack builds.
</blockquote></details>

Seems reasonable to me, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039)*

**Comment by mrtracy** [comment]
Review status: 49 of 52 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ui/src/components/listLink.tsx, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXYDPPZ7Bd0UDwuWL:-Kklpah-yhEIZz4UM-EC:b7rqve) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/components/listLink.tsx#L26)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

This one is different from the other two; the other two use
```
  // Magic to add react router to the context.
  // See https://github.com/ReactTraining/react-router/issues/975
  static contextTypes = {
    router: PropTypes.object.isRequired,
  }
```

Can we at least make them all the same?

Also, it looks like the referenced issue now talks about [`withRouter`](https://reacttraining.com/react-router/web/api/withRouter), which might be what we want here instead?
</blockquote></details>

Looked into withRouter, the typings aren't great for 3.x, but upgrading to 4.x isn't trivial.  Added a TODO comment at all three locations where contextTypes are used.

---

*[pkg/ui/src/containers/nodes.tsx, line 52 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KklUhB6KjFvR1i21NPn:-Kklpivj3eWhw0b_WpzA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebe4be9310c8ffbe238d16cd06ead0c09001e79/pkg/ui/src/containers/nodes.tsx#L52)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

given the `isRequired` above, I think this probably shouldn't be optional here.
</blockquote></details>

Done.

---

*[pkg/ui/src/containers/databases/databases.tsx, line 38 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KklUmaadj-tbq7fLVun:-KklpjNCAdByOavZDb9o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebe4be9310c8ffbe238d16cd06ead0c09001e79/pkg/ui/src/containers/databases/databases.tsx#L38)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

given the `isRequired` above, I think this probably shouldn't be optional here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/ui/src/components/listLink.tsx, line 26 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039#-KkYXYDPPZ7Bd0UDwuWL:-KklsCzIqn6fC-rQeuyo:b-6mhyx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6d5da3c415c853da53741ac33ef057d4d4e3541/pkg/ui/src/components/listLink.tsx#L26)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

Looked into withRouter, the typings aren't great for 3.x, but upgrading to 4.x isn't trivial.  Added a TODO comment at all three locations where contextTypes are used.
</blockquote></details>

Great, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16039)*
